Fill out the how the web works learn content (#38109)
mdn/content: files/en-us/learn_web_development/getting_started/web_standards/how_the_web_works/index.md

![Two circles representing client and server. An arrow labelled request is going from client to server, and an arrow labelled responses is going from server to client](simple-client-server.png)

- Clients are the typical web user's internet-connected devices (for example, your computer connected to your Wi-Fi, or your phone connected to your mobile network) and web-accessing software available on those devices (usually a web browser like Firefox or Chrome).

> - Servers are computers that store webpages, sites, or apps. When a client device wants to access a webpage, a copy of the webpage code is downloaded from the server onto the client machine to rendered by the browser and displayed to the user.

Reviewer: ```suggestion
- Servers are computers that store webpages, sites, or apps. When a client device (a visitor) wants to access a webpage, a copy of the webpage code is downloaded from the server onto the client machine to rendered by the browser and displayed to the user.
```

Author: I didn't use this suggestion. I think it confuses more than it helps, as we've not used "visitor" anywhere else, and the relationship is pretty clear if you read both bullets together.

Reviewer: no problem

## The other parts of the toolbox

The client and server we've described above don't tell the whole story. There are many other parts involved, and we'll describe them below.

For now, let's imagine that the internet is a road. On one end of the road is the client, which is like your house. On the other end of the road is the server, which is like a shop you want to buy something from.

![A black-and-white photo of a person crossing a road at a crosswalk](road.jpg)

> In addition to the client and the server, we also need to know about the other things involved in navigating to a web page:

Reviewer: ```suggestion
In order for data to get back and forth, we need the following things:
```

- **TCP/IP**: **Transmission Control Protocol** and **Internet Protocol** (TCP/IP) are communication protocols that define how data should travel across the internet. This is like the transport mechanisms that let you place an order, go to the shop, and buy your goods. In our example, this is like a car or a bike (or however else you might travel along the road).
- **DNS**: The **Domain Name System** (DNS) is like an address book for websites. When you type a web address in your browser, the browser looks at the DNS to find the website's IP address — the actual address the server is located at — before it can retrieve the website (see [DNS explained](#dns_explained) below for more information). The browser needs to find out which server the website lives on, so it can send HTTP messages to the right place (see below). This is like looking up the address of the shop before you visit it.
- **HTTP**: **Hypertext Transfer Protocol** (HTTP) is an application {{Glossary("Protocol", "protocol")}} that defines a language for clients and servers to speak to each other. This is like the language you use to order your goods. See [HTTP basics](#http_basics) below.

> - **Component files**: A website is made up of many different files, which are like the different goods you buy from the shop. These files come in two main types:

Reviewer: What do you think about this? Components is a little loaded:

```suggestion
- **Resources**: A website is made up of many different files, which are like the different goods you buy from the shop. These files come in two main types:
```

Reviewer: Sounds good 👍🏻

Websites can be reached directly via their IP addresses. You can use a DNS lookup tool to find the IP address of a website.

> [!CALLOUT]
>
> **Try it out**
>
> 1. Go to the [NsLookup.io DNS lookup tool](https://www.nslookup.io/website-to-ip-lookup/), type in `developer.mozilla.org`, and press the button.
> 2. In the results screen, copy the IP Address (the IPv4 address) to your system clipboard.
> 3. Open a new browser tab, paste the IP Address into the address bar and press <kbd>Enter</kbd>/<kbd>Return</kbd>. You should see MDN load up, proving that the IP address points to it.

## Packets explained

Earlier we used the term "packets" to describe the format in which the data is transferred between the client and server. What do we mean here?

Basically, when data is sent across the web, it is sent in thousands of small chunks. There are multiple reasons why data is sent in small packets, but most significantly:

- They are sometimes dropped or corrupted and, when this happens, it's quicker and easier to replace small chunks than entire files.

> - Additionally, the packets can be routed along different paths, making the exchange faster and allowing many different users to download the same website at the same time. If each website was sent as a single big chunk, only one user could download it at a time, which obviously would make the web very inefficient and not much fun to use.

Reviewer: ```suggestion
- Additionally, the packets can be routed along different paths, making the exchange faster and allowing many different users to download the same website at the same time. If each website was sent as a single big chunk, only one user could download it at a time, which would make the web very inefficient and not much fun to use.
```

HTTP uses a simple language of verbs to perform actions such as making requests (see [HTTP request methods](/en-US/docs/Web/HTTP/Methods)). The HTTP [`GET`](/en-US/docs/Web/HTTP/Methods/GET) method is the one normally used to make HTTP requests of the type described above. For example, a request for the MDN home page might look like this:

```http
GET / HTTP/2

Host: https://developer.mozilla.org/en-US/
```

The response sent by the server might looks something like this:

```http
HTTP/2 200

date: Tue, 11 Feb 2025 11:13:30 GMT
expires: Tue, 11 Feb 2025 11:40:01 GMT
server: Google frontend
last-modified: Tue, 11 Feb 2025 00:49:32 GMT
etag: "65f26b7f6463e2347f4e5a7a2adcee54"
content-length: 45227
content-type: text/html

<!doctype html> ... (the 45227 bytes of the requested web page HTML)
```

The full response is more complex than this, but we have omitted most of it for brevity. The main parts are as follows:

- `HTTP/2 200`
  - : The version of HTTP that the server is using to send the response, in this case HTTP/2, followed by a [status code](/en-US/docs/Web/HTTP/Status) indicating whether the request was successful. `200` indicates success.
- `date`, `expires`, etc.

>   - : [HTTP headers](/en-US/docs/Web/HTTP/Headers) containing additional information about the reponse (note that requests can have headers too), which provide extra information and/or modify its behavior.

Reviewer: ```suggestion
  - : [HTTP headers](/en-US/docs/Web/HTTP/Headers) containing additional information about the response (note that requests can have headers too), which provide extra information and/or modify its behavior.
```

- `<!doctype html>`, etc.
  - : The response body, which in this case contains the MDN homepage's HTML document.

> [!NOTE]
> See the MDN [HTTP reference](/en-US/docs/Web/HTTP) for a lot more detail on HTTP, if you are curious. [An overview of HTTP](/en-US/docs/Web/HTTP/Overview) is a good place to start.

### Other status codes

> Above, we met the `200` [status code](/en-US/docs/Web/HTTP/Status), which indicates that the HTTP request was successful. There are many other status code that you'll come across; the most common ones are:

Reviewer: ```suggestion
Above, we met the `200` [status code](/en-US/docs/Web/HTTP/Status), which indicates that the HTTP request was successful. There are many HTTP status codes that have specific meaning and uses, but you will probably only see a few very common ones:
```

Author: I thought about this, and ended up with

> Above, we met the `200` [status code](/en-US/docs/Web/HTTP/Status), which indicates that the HTTP request was successful. There are many HTTP status codes with specific meanings and uses, but you will only commonly see a few:

WDYT?

Reviewer: yep, also good 👍🏻

- `301`

>   - : The requested resource has been permanently moved to a new location, which is provided in the response.

Reviewer: Want to add some hints why you'll see them?

```suggestion
  - : The requested resource has been permanently moved to a new location, which is provided in the response. This is used for redirecting content when it's moved.
```

- `400`

>   - : The server cannot or will not process the request due to a problem with the request itself.

Reviewer: ```suggestion
  - : The server can't process the request. This usually happens if the request format is something the server can't process, or the request was badly-formed, in general.
```

Author: I ended up with

> The server can't process the request. This usually happens when the request isn't in a format the server understands, or has errors in it.

I think "badly-formed" might confuse people, as it is quite a specific validation term.

Reviewer: looking good

- `403`

>   - : The server will not give the client access to the requested resource because it does not have the necessary permissions to access it.

Reviewer: ```suggestion
  - : The server will not give the client access to the requested resource. This usually happens when the server knows who the client is, but they don't have permissions to access something.
```

Author: Tweaked it to

> The server will not give the client access to the requested resource. This usually happens when the server knows who the client is, but they don't have permissions to access something.

Reviewer: also fine, tnx

- `404`

>   - : The server cannot find the requested resource.

Reviewer: ```suggestion
  - : The server cannot find the requested resource. This is very common if the URL is wrong or if content is deleted without redirects in place.
```

Author: Ended up with

> The server cannot find the requested resource. This status is commonly returned if the URL is wrong or if content is deleted without putting a redirect in place.

- `503`

>   - : The request cannot be handled due to a problem with the server.

Reviewer: ```suggestion
  - : The request cannot be handled due to a problem with the server. This is common when servers are offline for maintenance, and it's expected to be temporary.
```

Reviewer: We could also say here, you can tell what type of problem it is by the first digit, or even copy:

https://github.com/mdn/content/blob/8dac6c62fc3cee2de82960d4dd9d9be16a3a1761/files/en-us/web/http/status/index.md?plain=1#L13-L17

Author: I don't really want to share this detail here; TMI. The reader can follow the link to the status codes page if they want more information.

## Components of a URL

Technically, web addresses that you type into the browser address bar form part of **Uniform Resource Locators** (**URLs**). URLs define the locations of unique resources on the internet.

A URL is a web address plus a protocol: for example, if you open a new tab in your browser, type in `developer.mozilla.org` into the address bar, and press <kbd>Enter</kbd>/<kbd>Return</kbd>, you will be redirected to a URL like the following one:

```plain
https://developer.mozilla.org/en-US/
```

The main parts of the URL are:

- `https`

>   - : The **protocol** being used to send the request. In this case, we are using [HTTPS](/en-US/docs/Glossary/HTTPS), which is a secure version of HTTP that stops bad people from stealing data while it is being transported. On the modern web, pretty much every server uses HTTPS, so if you don't include it explicitly, the browser assumes that is what you are using and adds it for you.

Reviewer: ```suggestion
  - : The **protocol** being used to send the request. In this case, we are using [HTTPS](/en-US/docs/Glossary/HTTPS), which is a secure version of HTTP that stops bad people from reading your data while it is being transported. On the modern web, pretty much every server uses HTTPS, so if you don't include it explicitly, the browser assumes that is what you are using and adds it for you.
```
